File: plugins/plugin-trajectory-logger/rust/src/export.rs
use crate::art_format::{group_trajectories, to_art_trajectory};
use crate::types::{Trajectory, TrajectoryGroup};
use std::fs;
use std::path::Path;

/// Write trajectories as ART JSONL to a file.
pub fn export_for_openpipe_art<P: AsRef<Path>>(
    dataset_name: &str,
    trajectories: &[Trajectory],
    out_path: Option<P>,
) -> std::io::Result<String> {
    let path = match out_path {
        Some(p) => p.as_ref().to_path_buf(),
        None => {
            let safe = dataset_name
                .chars()
                .map(|c| {
                    if c.is_ascii_alphanumeric() || c == '.' || c == '_' || c == '-' {
                        c
                    } else {
                        '_'
                    }
                })
                .collect::<String>();
            std::env::current_dir()?.join(format!("{}.trajectories.art.jsonl", safe))
        }
    };

    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }

    let mut lines = String::new();
    for t in trajectories {
        let art = to_art_trajectory(t);
        lines.push_str(&serde_json::to_string(&art).unwrap_or_else(|_| "{}".to_string()));
        lines.push('\n');
    }

    fs::write(&path, lines)?;
    Ok(path.to_string_lossy().to_string())
}

/// Write GRPO groups (JSON) to a file.
pub fn export_grouped_for_grpo<P: AsRef<Path>>(
    dataset_name: &str,
    trajectories: &[Trajectory],
    now_ms: i64,
    out_path: Option<P>,
) -> std::io::Result<String> {
    let path = match out_path {
        Some(p) => p.as_ref().to_path_buf(),
        None => {
            let safe = dataset_name
                .chars()
                .map(|c| {
                    if c.is_ascii_alphanumeric() || c == '.' || c == '_' || c == '-' {
                        c
                    } else {
                        '_'
                    }
                })
                .collect::<String>();
            std::env::current_dir()?.join(format!("{}.trajectories.grpo.groups.json", safe))
        }
    };

    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }

    let groups: Vec<TrajectoryGroup> = group_trajectories(trajectories, now_ms);
    let json = serde_json::to_string_pretty(&groups).unwrap_or_else(|_| "[]".to_string());
    fs::write(&path, format!("{}\n", json))?;
    Ok(path.to_string_lossy().to_string())
}

File: plugins/plugin-trajectory-logger/python/elizaos_plugin_trajectory_logger/export.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

from elizaos_plugin_trajectory_logger.art_format import group_trajectories, to_art_trajectory
from elizaos_plugin_trajectory_logger.types import Trajectory


@dataclass(frozen=True)
class ExportOptions:
    dataset_name: str
    trajectories: list[Trajectory]
    max_trajectories: int | None = None
    output_path: str | None = None
    output_dir: str | None = None


@dataclass(frozen=True)
class ExportResult:
    success: bool
    trajectories_exported: int
    dataset_url: str | None = None
    error: str | None = None


def export_for_openpipe_art(options: ExportOptions) -> ExportResult:
    out_path = _resolve_output_path(options, "trajectories.art.jsonl")
    trajs = (
        options.trajectories[: options.max_trajectories]
        if options.max_trajectories
        else options.trajectories
    )

    lines = (
        "\n".join(json.dumps(to_art_trajectory(t).model_dump(mode="json")) for t in trajs) + "\n"
    )
    _write_text(out_path, lines)

    return ExportResult(
        success=True,
        trajectories_exported=len(trajs),
        dataset_url=str(out_path),
    )


def export_grouped_for_grpo(options: ExportOptions) -> ExportResult:
    out_path = _resolve_output_path(options, "trajectories.grpo.groups.json")
    trajs = (
        options.trajectories[: options.max_trajectories]
        if options.max_trajectories
        else options.trajectories
    )

    groups = group_trajectories(trajs)
    _write_text(out_path, json.dumps([g.model_dump(mode="json") for g in groups], indent=2) + "\n")

    return ExportResult(
        success=True,
        trajectories_exported=len(trajs),
        dataset_url=str(out_path),
    )


def _resolve_output_path(options: ExportOptions, filename: str) -> Path:
    if options.output_path:
        return Path(options.output_path)
    if options.output_dir:
        return Path(options.output_dir) / filename
    safe = "".join(ch if ch.isalnum() or ch in "._-" else "_" for ch in options.dataset_name)
    return Path.cwd() / f"{safe}.{filename}"


def _write_text(path: Path, content: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content, encoding="utf-8")
